src/darwin/Framework/CHIP/EndpointDescription/MTRAccessGrant.mm

```diff
@@ -0,0 +1,131 @@
+        }
+        return nil;
+    }
+
+    return [self initWithSubject:[nodeID copy] privilege:privilege authenticationMode:MTRAccessControlEntryAuthModeCASE];
+}
+
+- (nullable instancetype)initWithCASEAuthenticatedTag:(NSNumber *)caseAuthenticatedTag privilege:(MTRAccessControlEntryPrivilege)privilege error:(NSError * __autoreleasing *)error
+{
+    auto value = caseAuthenticatedTag.unsignedLongLongValue;
+    if (!CanCastTo<CASEAuthTag>(value)) {
+        MTR_LOG_ERROR("MTRAccessGrant provided too-large CAT value: 0x%llx", value);
+        if (error) {
+            *error = [MTRError errorForCHIPErrorCode:CHIP_ERROR_INVALID_ARGUMENT];
+        }
+        return nil;
+    }
+
+    CASEAuthTag tag = static_cast<CASEAuthTag>(value);
+    if (!IsValidCASEAuthTag(tag)) {
+        MTR_LOG_ERROR("MTRAccessGrant provided invalid CAT value: 0x%" PRIx32, tag);
+        if (error) {
+            *error = [MTRError errorForCHIPErrorCode:CHIP_ERROR_INVALID_ARGUMENT];
+        }
+        return nil;
+    }
+
+    return [self initWithSubject:@(NodeIdFromCASEAuthTag(tag)) privilege:privilege authenticationMode:MTRAccessControlEntryAuthModeCASE];
+}
+
+- (nullable instancetype)initWithGroupID:(NSNumber *)groupID privilege:(MTRAccessControlEntryPrivilege)privilege error:(NSError * __autoreleasing *)error
+{
+    auto value = groupID.unsignedLongLongValue;
+    if (!CanCastTo<GroupId>(value)) {
+        MTR_LOG_ERROR("MTRAccessGrant provided too-large group id: 0x%llx", value);
+        if (error) {
+            *error = [MTRError errorForCHIPErrorCode:CHIP_ERROR_INVALID_ARGUMENT];
+        }
+        return nil;
+    }
+
+    GroupId id = static_cast<GroupId>(value);
+    if (!IsValidGroupId(id)) {
+        MTR_LOG_ERROR("MTRAccessGrant provided invalid group id: 0x%" PRIx32, id);
+        if (error) {
+            *error = [MTRError errorForCHIPErrorCode:CHIP_ERROR_INVALID_ARGUMENT];
+        }
+        return nil;
+    }
+
+    return [self initWithSubject:@(NodeIdFromGroupId(id)) privilege:privilege authenticationMode:MTRAccessControlEntryAuthModeGroup];
+}
+
+- (instancetype)initForAllNodesWithPrivilege:(MTRAccessControlEntryPrivilege)privilege
+{
+    return [self initWithSubject:nil privilege:privilege authenticationMode:MTRAccessControlEntryAuthModeCASE];
+}
+
+// initWithSubject assumes that the subject has already been validated and, if
+// needed, copied from the input.
+- (instancetype)initWithSubject:(nullable NSNumber *)subject privilege:(MTRAccessControlEntryPrivilege)privilege authenticationMode:(MTRAccessControlEntryAuthMode)authenticationMode
+{
+    if (!(self = [super init])) {
+        return nil;
+    }
+
+    _subjectID = subject;
+    _grantedPrivilege = privilege;
+    _authenticationMode = authenticationMode;
+    return self;
+}
+
+- (id)copyWithZone:(NSZone *)zone
+{
+    return [[MTRAccessGrant alloc] initWithSubject:[_subjectID copy] privilege:_grantedPrivilege authenticationMode:_authenticationMode];
```

Comment:
It looks like this class is immutable, so this can just return self?

Reply:
Yes, I can do that.  As long as people don't start relying on that pointer identity....